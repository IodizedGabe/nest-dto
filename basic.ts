export * from './src/flavors/basic';
