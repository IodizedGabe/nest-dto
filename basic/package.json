{
    "main": "../dist/flavors/basic.js",
    "types": "../dist/flavors/basic.d.ts"
}
