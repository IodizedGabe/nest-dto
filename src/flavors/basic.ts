/* Defines a basic set of decorators.
 */
import { createBuilder } from '../builder';
import {
    ArrayOptions,
    BooleanOptions,
    DateOptions,
    DateStringOptions,
    EnumOptions,
    Flavor,
    IntegerOptions,
    NestedOptions,
    NumberOptions,
    StringOptions,
    UUIDOptions,
} from '../interfaces';
import {
    TransformerOptions,
    ValidatorOptions,
    withTransformer,
    withValidator,
} from '../mixins';
import {
    IsArrayRecipe,
    IsBooleanRecipe,
    IsDateRecipe,
    IsDateStringRecipe,
    IsEnumRecipe,
    IsIntegerRecipe,
    IsNestedRecipe,
    IsNumberRecipe,
    IsStringRecipe,
    IsUUIDRecipe,
} from '../recipes';

export interface FlavorOptions extends TransformerOptions, ValidatorOptions {}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function createBuilderWithMixins<Options>() {
    return withTransformer(
        withValidator(
            createBuilder<Options>(),
        ),
    );
}

export const flavor: Flavor<FlavorOptions> = {
    IsArray: IsArrayRecipe(createBuilderWithMixins<FlavorOptions & ArrayOptions>()),
    IsBoolean: IsBooleanRecipe(createBuilderWithMixins<FlavorOptions & BooleanOptions>()),
    IsDate: IsDateRecipe(createBuilderWithMixins<FlavorOptions & DateOptions>()),
    IsDateString: IsDateStringRecipe(createBuilderWithMixins<FlavorOptions & DateStringOptions>()),
    IsEnum: IsEnumRecipe(createBuilderWithMixins<FlavorOptions & EnumOptions>()),
    IsInteger: IsIntegerRecipe(createBuilderWithMixins<FlavorOptions & IntegerOptions>()),
    IsNested: IsNestedRecipe(createBuilderWithMixins<FlavorOptions & NestedOptions>()),
    IsNumber: IsNumberRecipe(createBuilderWithMixins<FlavorOptions & NumberOptions>()),
    IsString: IsStringRecipe(createBuilderWithMixins<FlavorOptions & StringOptions>()),
    IsUUID: IsUUIDRecipe(createBuilderWithMixins<FlavorOptions & UUIDOptions>()),
};

export type BasicFlavor = typeof flavor;
