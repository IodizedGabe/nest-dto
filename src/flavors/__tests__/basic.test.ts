import { plainToClass } from 'class-transformer';
import { validate } from 'class-validator';

import * as basic from '../basic';
import { INPUT, createFixtures } from './fixtures';

describe('flavors.basic', () => {
    const Example = createFixtures(basic);

    it('validates required fields', async () => {
        const obj = plainToClass(Example, {});

        // we do not expect any data to be transformed
        expect(Object.keys(obj)).toHaveLength(0);
        expect(obj).toMatchSnapshot();

        const errors = await validate(obj);
        // we expect an error for every required field (but not the optional ones)
        expect(errors).toHaveLength(11);
        expect(errors).toMatchSnapshot();
    });
    it('transforms input data', async () => {
        const obj = plainToClass(Example, INPUT);

        // expect all data to be transformed
        expect(Object.keys(obj)).toHaveLength(33);
        expect(obj).toMatchObject({
            requiredObjectArrayValue: [{
                requiredStringValue: 'nested',
            }],
            requiredPrimitiveArrayValue: [
                42,
            ],
            requiredNestedValue: {
                requiredStringValue: 'nested',
            },
            optionalObjectArrayValue: [{
                optionalStringValue: 'nested',
            }],
            optionalPrimitiveArrayValue: [
                42,
            ],
            optionalNestedValue: {
                optionalStringValue: 'nested',
            },
            nullableObjectArrayValue: [{
                nullableStringValue: 'nested',
            }],
            nullablePrimitiveArrayValue: [
                42,
            ],
            nullableNestedValue: {
                nullableStringValue: 'nested',
            },
        });

        const errors = await validate(obj);
        // we expect no errors
        expect(errors).toHaveLength(0);
        expect(errors).toMatchSnapshot();
    });
    it('does not transform input data if extraneous values are excluded', async () => {
        const obj = plainToClass(Example, INPUT, {
            excludeExtraneousValues: true,
        });

        // expect no data to be transformed (b/c @Expose() is not included)
        expect(Object.keys(obj)).toHaveLength(0);

        const errors = await validate(obj);
        // we expect an error for every required field (but not the optional ones)
        expect(errors).toHaveLength(11);
        expect(errors).toMatchSnapshot();
    });
});
