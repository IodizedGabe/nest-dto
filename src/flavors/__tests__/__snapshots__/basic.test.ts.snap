// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flavors.basic does not transform input data if extraneous values are excluded 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredObjectArrayValue must be an array",
      "isDefined": "requiredObjectArrayValue should not be null or undefined",
    },
    "property": "requiredObjectArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredPrimitiveArrayValue must be an array",
      "isDefined": "requiredPrimitiveArrayValue should not be null or undefined",
    },
    "property": "requiredPrimitiveArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "requiredBooleanValue must be a boolean value",
      "isDefined": "requiredBooleanValue should not be null or undefined",
    },
    "property": "requiredBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredDateStringValue should not be null or undefined",
      "isIso8601": "requiredDateStringValue must be a valid ISO 8601 date string",
    },
    "property": "requiredDateStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDate": "requiredDateValue must be a Date instance",
      "isDefined": "requiredDateValue should not be null or undefined",
    },
    "property": "requiredDateValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredEnumValue should not be null or undefined",
      "isEnum": "requiredEnumValue must be a valid enum value",
    },
    "property": "requiredEnumValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredIntegerValue should not be null or undefined",
      "isInt": "requiredIntegerValue must be an integer number",
    },
    "property": "requiredIntegerValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNestedValue should not be null or undefined",
    },
    "property": "requiredNestedValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNumberValue should not be null or undefined",
      "isNumber": "requiredNumberValue must be a number conforming to the specified constraints",
    },
    "property": "requiredNumberValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredStringValue should not be null or undefined",
      "isString": "requiredStringValue must be a string",
    },
    "property": "requiredStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredUUIDValue should not be null or undefined",
      "isUuid": "requiredUUIDValue must be an UUID",
    },
    "property": "requiredUUIDValue",
    "target": Example {},
    "value": undefined,
  },
]
`;

exports[`flavors.basic transforms input data 1`] = `Array []`;

exports[`flavors.basic validates required fields 1`] = `Example {}`;

exports[`flavors.basic validates required fields 2`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredObjectArrayValue must be an array",
      "isDefined": "requiredObjectArrayValue should not be null or undefined",
    },
    "property": "requiredObjectArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredPrimitiveArrayValue must be an array",
      "isDefined": "requiredPrimitiveArrayValue should not be null or undefined",
    },
    "property": "requiredPrimitiveArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "requiredBooleanValue must be a boolean value",
      "isDefined": "requiredBooleanValue should not be null or undefined",
    },
    "property": "requiredBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredDateStringValue should not be null or undefined",
      "isIso8601": "requiredDateStringValue must be a valid ISO 8601 date string",
    },
    "property": "requiredDateStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDate": "requiredDateValue must be a Date instance",
      "isDefined": "requiredDateValue should not be null or undefined",
    },
    "property": "requiredDateValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredEnumValue should not be null or undefined",
      "isEnum": "requiredEnumValue must be a valid enum value",
    },
    "property": "requiredEnumValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredIntegerValue should not be null or undefined",
      "isInt": "requiredIntegerValue must be an integer number",
    },
    "property": "requiredIntegerValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNestedValue should not be null or undefined",
    },
    "property": "requiredNestedValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNumberValue should not be null or undefined",
      "isNumber": "requiredNumberValue must be a number conforming to the specified constraints",
    },
    "property": "requiredNumberValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredStringValue should not be null or undefined",
      "isString": "requiredStringValue must be a string",
    },
    "property": "requiredStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredUUIDValue should not be null or undefined",
      "isUuid": "requiredUUIDValue must be an UUID",
    },
    "property": "requiredUUIDValue",
    "target": Example {},
    "value": undefined,
  },
]
`;
